source/lib/src/cuda/prod_env_mat.cu

```diff
@@ -0,0 +1,514 @@
+#include "gpu_cuda.h"
+#include "prod_env_mat.h"
+#include <cub/block/block_load.cuh>
+#include <cub/block/block_store.cuh>
+#include <cub/block/block_radix_sort.cuh>
+
+// common part of prod_env_mat
+template <
+    typename    Key,
+    int         BLOCK_THREADS,
+    int         ITEMS_PER_THREAD>
+__launch_bounds__ (BLOCK_THREADS)
+__global__ void BlockSortKernel(
+    Key * d_in,
+    Key * d_out)                // Tile of output
+{   
+  enum { TILE_SIZE = BLOCK_THREADS * ITEMS_PER_THREAD };
+  // Specialize BlockLoad type for our thread block (uses warp-striped loads for coalescing, then transposes in shared memory to a blocked arrangement)
+  typedef cub::BlockLoad<Key, BLOCK_THREADS, ITEMS_PER_THREAD, cub::BLOCK_LOAD_WARP_TRANSPOSE> BlockLoadT;
+  // Specialize BlockRadixSort type for our thread block
+  typedef cub::BlockRadixSort<Key, BLOCK_THREADS, ITEMS_PER_THREAD> BlockRadixSortT;
+  // Shared memory
+  __shared__ union TempStorage
+  {
+    typename BlockLoadT::TempStorage        load;
+    typename BlockRadixSortT::TempStorage   sort;
+  } temp_storage;
+  // Per-thread tile items
+  Key items[ITEMS_PER_THREAD];
+  // Our current block's offset
+  int block_offset = blockIdx.x * TILE_SIZE;
+  // Load items into a blocked arrangement
+  BlockLoadT(temp_storage.load).Load(d_in + block_offset, items);
+  // Barrier for smem reuse
+  __syncthreads();
+  // Sort keys
+  BlockRadixSortT(temp_storage.sort).SortBlockedToStriped(items);
+  // Store output in striped fashion
+  cub::StoreDirectStriped<BLOCK_THREADS>(threadIdx.x, d_out + block_offset, items);
+}
+
+template<typename FPTYPE>
+__device__ inline FPTYPE dev_dot(
+    FPTYPE * arr1, 
+    FPTYPE * arr2) 
+{
+  return arr1[0] * arr2[0] + arr1[1] * arr2[1] + arr1[2] * arr2[2];
+}
+
+template<typename FPTYPE>
+__device__ inline void spline5_switch(
```

Comment:
Place this function in `switcher.h`?

Reply:
It's a device function, nvcc compiler is needed, so better to leave it here.

Reply:
Conditional compiling solve the problem